---
title: "Using the MethylSeqR package"
author: "Hailey Zimmerman and Jonathon T. Hill, PhD"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    css: "bioconductor2.css"
vignette: >
  %\VignetteIndexEntry{Using the MethylSeqR package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---
  
```{r, include=FALSE}
library(knitr, quietly=TRUE)
library(MethylSeqR, quietly=TRUE)
# library(Biostrings, quietly=TRUE)
opts_chunk$set(tidy=TRUE, tidy.opts=list(width.cutoff=70))
```

# Introduction

The `MethylSeqR` package provides basic functions for importing and 
working with nanopore methylation sequencing data.

This vignette will walk you through a typical workflow using CH3 files created 
by Wasatch Biolabs (https://www.wasatchbiolabs.com/). These files contain results 
from a nanopore sequencing run. A CH3 file will be created for each sample within 
a run, and is meant as an intermediate existing to compress large amounts of 
data into a usable form. 

It's most helpful to store all the sample CH3 files from a run you want to 
analyze within a single directory to get started. 

# Loading Data
The first step of a workflow will be to upload data from a directory holding 
sample sequencing data in the form of CH3 files. These files are typically 
created by Wasatch Biolabs.

Once data is obtained in this format, the first step is done using the 
`make_ch3_db()` function. This function directly imports a directory of CH3 
files, and creates a database in the form of a local file.

## make_pos_db
`make_ch3_db()` requires 2 arguments. First, A string representing the directory
path containing CH3 files (`ch3_files`), and second, a string representing the 
path where the positions database will be created (`db_name`). 

4 additional special arguments exist if needed. `chrom` allows a list representation 
of all chromosomes wanted to be included in the positional data, not limited to 
the human genome. `min_read_length` is a numeric value 
representing the minimum read length. Only reads meeting or exceeding this length
will be processed. Default is 50. `min_call_prob` is a numeric value representing the minimum call 
probability threshold.  Only calls with a probability greater than or equal to 
this value will be included. Default is 0.9. `min_base_qual` is a numeric value 
representing the minimum base quality. Only reads with quality scores at or above 
this threshold will be included. Default is 10. `flag` An optional numeric value 
specifying a flag-based filter for the data.If NULL, no flag filtering is applied.
And lastly, `chr_prefix`is a boolean value stating whether or not to keep the chr prefix 
in the chromosome column. Default is TRUE, and "chr" will be kept in front of every
chr number.

Here is an example:

```{r}
# Set up the file path for the test data located in inst/test_data/
ch3_files <- system.file("TEST_DATA", package = "MethylSeqR")
ch3_db <- tempfile("example_db")

# Run the function with the example data
make_pos_db(ch3_files, ch3_db)
```


After this step, a database is made and the raw calls table is included. A user 
may continue their analysis in any direction they want. They may run a quality control 
(`run_qc()`), summarize data by positions (`summarize_positions()`),
regions (`summarize_regions()`), or windows (`summarize_windows()`). 
At any point in the analysis, a user can export a table in the database by calling
`export_tables()`, or use `get_table()`to import as a tibble into the local environment.
All these functions will be explained below.

## summarize_positions
This function summarizes methylation data (from the calls table) by each unique 
position in the genome.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")

# Summarize positions using annotation table
summarize_positions(ch3_db)
```

A user must provide the path to the database with the argument `ch3_db`, but optionally
can include the modificaton type they want (m, h, or mh) with the argument `mod_type`,
the unique chromosomes with `chrs`, or the minimum number of calls required to 
include a position in the summary with `min_num_calls` (default is 1).

## summarize_regions
This function summarizes methylation data based on specified regions defined in 
a BED, TSV, or CSV file. It performs a join operation between the methylation 
data and the regions specified in the annotation file, allowing for different types of joins.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")

region_bed = system.file("Islands_hg38_test.csv", package = "MethylSeqR")
# Summarize Regions using annotation table
summarize_regions(ch3_db, region_bed)
```

There are 2 required arguments. `ch3_db` is a list containing the database file 
path. This should be a string representing a path to the database, or a valid 
"ch3_db" class object. `region_file` asks for a string representing the 
path to the BED or CSV file that contains the region annotations. Optionally,
a user can include the modificaton type they want (m, h, or mh) with the argument 
`mod_type`, the unique chromosomes with `chrs`, or the minimum number of calls required to 
include a region in the summary with `min_num_calls` (default is 1).

## summarize_windows
This function summarizes methylation data from a ch3_db database by creating 
sliding windows over the specified genomic regions. It allows for the adjustment 
of window size and step size to control the granularity of the summarization.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
 
# Calculate windows
summarize_windows(ch3_db, window_size = 100, step_size = 100)
```

There is only 1 required arguments. `ch3_db` is a list containing the database 
file path. This should be a string representing a path to the database, or a valid
"ch3_db" class object. `window_size` (integer) specifys the size of the sliding 
window in base pairs. Default is 1000. `step_size` (integer) specifies the 
number of base pairs to step forward with each window. Default is 10. 
Optionally, a user can include the modificaton type they want (m, h, or mh) with
the argument `mod_type`, the unique chromosomes with `chrs`, or the minimum number 
of calls required to include a window in the summary with `min_num_calls` (default is 1).
`overwrite` is a logical indicating whether to overwrite the existing "windows" 
table if it exists. Default is TRUE.

# Differential Methylation
## calc_mod_diff
This function calculates differential methylation between specified case and control groups using various statistical methods. 

Differential methylation can be calculated from <b><i>positional</i></b>, <b><i>regional</i></b>, or <b><i>window</i></b> methylation data.

Calling this function creates a `meth_diff` table within the database, where differential methylation data can be found. As of our current model, calling `calc_mod_diff` again on a different set of data will overwrite previous data in the `meth_diff` table.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
  
# Get methylation statistics for the 'positions' call type without plotting
calc_mod_diff(ch3_db = ch3_db,
              call_type = "positions",
              cases = c("chr21_5xSample_BloodA", "chr21_5xSample_BloodE", "chr21_5xSample_BloodT"),
              controls = c("chr21_5xSample_SpermA", "chr21_5xSample_SpermB", "chr21_5xSample_SpermD"))
``` 

# Quality Control 

A variety of quality control functions are available to visually assess 
methylation data. 

## get_cov_stats
This function calculates and optionally plots statistics for coverage data from
methylation sequencing experiments. It can handle both positional and regional
methylation data.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
 
# Get coverage statistics for the 'positions' call type without plotting
get_cov_stats(ch3_db = ch3_db, call_type = "positions")
# Get coverage statistics for the 'positions' call type with plotting
get_cov_stats(ch3_db = ch3_db, call_type = "positions", plot = TRUE)
``` 

## get_mod_stats
This function retrieves and calculates methylation statistics (mean methylation 
fractions) from a specified table in the ch3 database. It can either return 
summary statistics or plot a histogram of the methylation values, depending on 
the user's preference. 

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
 
# Get methylation statistics for the 'positions' call type without plotting
get_mod_stats(ch3_db = ch3_db, call_type = "positions")
get_mod_stats(ch3_db = ch3_db, call_type = "positions", plot = TRUE)
``` 

## cor_modseq
This function calculates and optionally plots a correlation matrix for 
methylation or other modification fraction data from genomic positions. 
It can handle both position-based and region-based calls and supports 
visualization using ggplot2.

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
 
# Run the correlation matrix function using the 'positions' call type and plot the results
cor_modseq(ch3_db = ch3_db, call_type = "positions", plot = TRUE)
``` 

## pca_modseq
This function performs Principal Component Analysis (PCA) on methylation data 
retrieved from the database.It aggregates the methylation fraction data 
based on the specified call type and prepares it for PCA analysis.
```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
 
# Calculate PCA
pca_modseq(ch3_db)
``` 


## qc_wrapper

Combining all functions above, the `qc_wrapper` function computes all QC checks
on methylation data at the same time. 

```{r}
# Specify the path to the database
ch3_db <- system.file("my_data.ch3.db", package = "MethylSeqR")
 
# Calculate windows
qc_wrapper(ch3_db)
``` 
 
# Retrieving and Exporting Tables
You can access data in your database with two different functions.

## get_table
To retrieve a single table into your local environment for further analysis, 
using `get_table` will transform it into a tibble and store it into a variable.
Arguments required includes the database, and name of the table to be retrieved.

## export_table
To export one or more tables into a CSV file, use `export_table`. This function 
takes three arguments, the database, which tables to export, and the path where 
the CSV files will be saved. 


# Conclusion
In this vignette, we have walked you through the basic functions in the 
`MethylSeqR` package. This work is a work in progress and we hope to 
improve its functionality. If you have any suggestions or 
requested features, please email Jonathon Hill at jhill@byu.edu.
